_ignored_source format flips for existing TSDB indices after #152481 (unexpected docvalues type NONE)

### Summary

Searches against existing TSDB indices fail with:

```
java.lang.IllegalStateException: unexpected docvalues type NONE for field '_ignored_source' (expected=BINARY). Re-index with correct docvalues type.
```

This was introduced by #152481, which graduated the columnar index mode feature flags to tech preview and, as part of that, removed the `IGNORED_SOURCE_AS_DOC_VALUES_FF` feature flag check from `IgnoredSourceFieldMapper#ignoredSourceFormat` **without bumping the gating index version**.

### Root cause

In `IgnoredSourceFieldMapper#ignoredSourceFormat(IndexSettings)`, the condition selecting `DOC_VALUES_IGNORED_SOURCE` previously required the feature flag to be enabled:

```java
// before #152481
if (IGNORED_SOURCE_AS_DOC_VALUES_FF.isEnabled()
    && indexCreatedVersion.onOrAfter(IndexVersions.IGNORED_SOURCE_AS_DOC_VALUES)
    && indexSettings.useTimeSeriesDocValuesFormat()) {
    return IgnoredSourceFormat.DOC_VALUES_IGNORED_SOURCE;
}
```

```java
// after #152481
if (indexCreatedVersion.onOrAfter(IndexVersions.IGNORED_SOURCE_AS_DOC_VALUES)
    && indexSettings.useTimeSeriesDocValuesFormat()) {
    return IgnoredSourceFormat.DOC_VALUES_IGNORED_SOURCE;
}
```

The gating index version `IGNORED_SOURCE_AS_DOC_VALUES` (`9_078_0_00`) was **not** changed.

`FeatureFlag` is disabled by default on release builds (only auto-enabled on snapshot builds, or via system property). So in environments where the flag was off, a TSDB index created at version >= `9_078_0_00` with the time-series doc-values codec resolved to `COALESCED_SINGLE_IGNORED_SOURCE` (stored-field-based), because the `&&` short-circuited on the disabled flag.

After #152481, the same already-created index now resolves to `DOC_VALUES_IGNORED_SOURCE`. The ignored-source format must remain fixed for the lifetime of an index. `_ignored_source` was physically written as stored fields, but the new code reads it as binary doc values — the synthetic source loader (`MultiValuedSortedBinaryDocValues.fromMultiValued` -> `DocValues.getBinary`) requests a BINARY docvalues field, finds NONE, and throws.

### Stack trace

```
Caused by: java.lang.IllegalStateException: unexpected docvalues type NONE for field '_ignored_source' (expected=BINARY). Re-index with correct docvalues type.
    at org.apache.lucene.core@10.5.0/org.apache.lucene.index.DocValues.checkField(DocValues.java:218)
    at org.apache.lucene.core@10.5.0/org.apache.lucene.index.DocValues.getBinary(DocValues.java:257)
    at org.elasticsearch.server@9.6.0/org.elasticsearch.index.fielddata.MultiValuedSortedBinaryDocValues.fromMultiValued(MultiValuedSortedBinaryDocValues.java:67)
    at org.elasticsearch.server@9.6.0/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.<init>(SourceLoader.java:218)
    at org.elasticsearch.server@9.6.0/org.elasticsearch.index.mapper.SourceLoader$Synthetic.leaf(SourceLoader.java:161)
    at org.elasticsearch.server@9.6.0/org.elasticsearch.search.fetch.FetchPhase$1.setNextReader(FetchPhase.java:351)
    ...
    at org.elasticsearch.server@9.6.0/org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:222)
    at org.elasticsearch.server@9.6.0/org.elasticsearch.search.SearchService$2.doRun(SearchService.java:1260)
```

### Impact

Existing TSDB indices created in the window where the feature flag was disabled but the index version was already >= `IGNORED_SOURCE_AS_DOC_VALUES` become unsearchable (synthetic source fetch fails) after upgrading to a build containing #152481. Observed in QA during a recent rollout.

### Suggested fix

Gate the `DOC_VALUES_IGNORED_SOURCE` branch on a **new, higher** `IndexVersion` (created after the fix ships) rather than restoring the flag. That way indices already created in the broken window continue to resolve to the stored-field-based `COALESCED_SINGLE_IGNORED_SOURCE` format they were written with, and only indices created after the fix use the doc-values format.

Introduced in: https://github.com/elastic/elasticsearch/pull/152481

--- a/server/src/main/java/org/elasticsearch/index/IndexVersions.java
+++ b/server/src/main/java/org/elasticsearch/index/IndexVersions.java
@@ -266,6 +266,7 @@ public class IndexVersions {
     public static final IndexVersion UPGRADE_TO_LUCENE_10_5_0 = def(9_104_0_00, Version.LUCENE_10_5_0);
     public static final IndexVersion DISK_BBQ_ES950_AUTO_CALIBRATE = def(9_105_0_00, Version.LUCENE_10_5_0);
     public static final IndexVersion TIME_SERIES_ES95_CODEC_DEFAULT = def(9_106_0_00, Version.LUCENE_10_5_0);
+    public static final IndexVersion IGNORED_SOURCE_AS_DOC_VALUES_NO_FF = def(9_107_0_00, Version.LUCENE_10_5_0);
 
     /*
      * STOP! READ THIS FIRST! No, really,

--- a/server/src/main/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapper.java
+++ b/server/src/main/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapper.java
@@ -611,7 +611,8 @@ public class IgnoredSourceFieldMapper extends MetadataFieldMapper {
     public static IgnoredSourceFormat ignoredSourceFormat(IndexSettings indexSettings) {
         IndexVersion indexCreatedVersion = indexSettings.getIndexVersionCreated();
         // we need TSDB doc values format to use binary doc values for ignored source, otherwise the source will be uncompressed
-        if (indexCreatedVersion.onOrAfter(IndexVersions.IGNORED_SOURCE_AS_DOC_VALUES) && indexSettings.useTimeSeriesDocValuesFormat()) {
+        if (indexCreatedVersion.onOrAfter(IndexVersions.IGNORED_SOURCE_AS_DOC_VALUES_NO_FF)
+            && indexSettings.useTimeSeriesDocValuesFormat()) {
             return IgnoredSourceFormat.DOC_VALUES_IGNORED_SOURCE;
         }
 

--- a/server/src/main/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapper.java
+++ b/server/src/main/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapper.java
@@ -13,6 +13,7 @@ import org.apache.lucene.document.StoredField;
 import org.apache.lucene.index.LeafReader;
 import org.apache.lucene.index.LeafReaderContext;
 import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.Build;
 import org.elasticsearch.ElasticsearchException;
 import org.elasticsearch.common.bytes.BytesArray;
 import org.elasticsearch.common.bytes.BytesReference;
@@ -611,8 +612,12 @@ public class IgnoredSourceFieldMapper extends MetadataFieldMapper {
     public static IgnoredSourceFormat ignoredSourceFormat(IndexSettings indexSettings) {
         IndexVersion indexCreatedVersion = indexSettings.getIndexVersionCreated();
         // we need TSDB doc values format to use binary doc values for ignored source, otherwise the source will be uncompressed
-        if (indexCreatedVersion.onOrAfter(IndexVersions.IGNORED_SOURCE_AS_DOC_VALUES_NO_FF)
-            && indexSettings.useTimeSeriesDocValuesFormat()) {
+
+        IndexVersion switchToDocValuesFormatVersion = Build.current().isSnapshot()
+            ? IndexVersions.IGNORED_SOURCE_AS_DOC_VALUES
+            : IndexVersions.IGNORED_SOURCE_AS_DOC_VALUES_NO_FF;
+
+        if (indexCreatedVersion.onOrAfter(switchToDocValuesFormatVersion) && indexSettings.useTimeSeriesDocValuesFormat()) {
             return IgnoredSourceFormat.DOC_VALUES_IGNORED_SOURCE;
         }
 

--- a/server/src/test/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapperTests.java
+++ b/server/src/test/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapperTests.java
@@ -2655,7 +2655,7 @@ public class IgnoredSourceFieldMapperTests extends MapperServiceTestCase {
         assertEquals(
             IgnoredSourceFieldMapper.IgnoredSourceFormat.DOC_VALUES_IGNORED_SOURCE,
             IgnoredSourceFieldMapper.ignoredSourceFormat(
-                createIndexSettings(IndexVersions.IGNORED_SOURCE_AS_DOC_VALUES, tsdbDocValuesFormatEnabled)
+                createIndexSettings(IndexVersions.IGNORED_SOURCE_AS_DOC_VALUES_NO_FF, tsdbDocValuesFormatEnabled)
             )
         );
 
@@ -2663,7 +2663,7 @@ public class IgnoredSourceFieldMapperTests extends MapperServiceTestCase {
         assertNotEquals(
             IgnoredSourceFieldMapper.IgnoredSourceFormat.DOC_VALUES_IGNORED_SOURCE,
             IgnoredSourceFieldMapper.ignoredSourceFormat(
-                createIndexSettings(IndexVersions.IGNORED_SOURCE_AS_DOC_VALUES, tsdbDocValuesFormatDisabled)
+                createIndexSettings(IndexVersions.IGNORED_SOURCE_AS_DOC_VALUES_NO_FF, tsdbDocValuesFormatDisabled)
             )
         );
 
@@ -2672,7 +2672,7 @@ public class IgnoredSourceFieldMapperTests extends MapperServiceTestCase {
             IgnoredSourceFieldMapper.IgnoredSourceFormat.DOC_VALUES_IGNORED_SOURCE,
             IgnoredSourceFieldMapper.ignoredSourceFormat(
                 createIndexSettings(
-                    IndexVersionUtils.getPreviousVersion(IndexVersions.IGNORED_SOURCE_AS_DOC_VALUES),
+                    IndexVersionUtils.getPreviousVersion(IndexVersions.IGNORED_SOURCE_AS_DOC_VALUES_NO_FF),
                     tsdbDocValuesFormatEnabled
                 )
             )

--- a/server/src/test/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapperTests.java
+++ b/server/src/test/java/org/elasticsearch/index/mapper/IgnoredSourceFieldMapperTests.java
@@ -10,11 +10,13 @@
 package org.elasticsearch.index.mapper;
 
 import org.apache.lucene.index.DirectoryReader;
+import org.elasticsearch.Build;
 import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.core.CheckedConsumer;
 import org.elasticsearch.core.Nullable;
 import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.IndexVersion;
 import org.elasticsearch.index.IndexVersions;
 import org.elasticsearch.search.lookup.SourceFilter;
 import org.elasticsearch.test.WildcardFieldMaskingReader;
@@ -2652,29 +2654,25 @@ public class IgnoredSourceFieldMapperTests extends MapperServiceTestCase {
             .build();
 
         // with TSDB doc values format enabled and matching index version, the doc values ignored source format should be used
+        IndexVersion switchToDocValuesFormatVersion = Build.current().isSnapshot()
+            ? IndexVersions.IGNORED_SOURCE_AS_DOC_VALUES
+            : IndexVersions.IGNORED_SOURCE_AS_DOC_VALUES_NO_FF;
         assertEquals(
             IgnoredSourceFieldMapper.IgnoredSourceFormat.DOC_VALUES_IGNORED_SOURCE,
-            IgnoredSourceFieldMapper.ignoredSourceFormat(
-                createIndexSettings(IndexVersions.IGNORED_SOURCE_AS_DOC_VALUES_NO_FF, tsdbDocValuesFormatEnabled)
-            )
+            IgnoredSourceFieldMapper.ignoredSourceFormat(createIndexSettings(switchToDocValuesFormatVersion, tsdbDocValuesFormatEnabled))
         );
 
         // index version alone is not enough, TSDB doc values setting must also be enabled.
         assertNotEquals(
             IgnoredSourceFieldMapper.IgnoredSourceFormat.DOC_VALUES_IGNORED_SOURCE,
-            IgnoredSourceFieldMapper.ignoredSourceFormat(
-                createIndexSettings(IndexVersions.IGNORED_SOURCE_AS_DOC_VALUES_NO_FF, tsdbDocValuesFormatDisabled)
-            )
+            IgnoredSourceFieldMapper.ignoredSourceFormat(createIndexSettings(switchToDocValuesFormatVersion, tsdbDocValuesFormatDisabled))
         );
 
         // TSDB doc values format is not enough, the index version must also match
         assertNotEquals(
             IgnoredSourceFieldMapper.IgnoredSourceFormat.DOC_VALUES_IGNORED_SOURCE,
             IgnoredSourceFieldMapper.ignoredSourceFormat(
-                createIndexSettings(
-                    IndexVersionUtils.getPreviousVersion(IndexVersions.IGNORED_SOURCE_AS_DOC_VALUES_NO_FF),
-                    tsdbDocValuesFormatEnabled
-                )
+                createIndexSettings(IndexVersionUtils.getPreviousVersion(switchToDocValuesFormatVersion), tsdbDocValuesFormatEnabled)
             )
         );
     }

